src/generators.py

```diff
@@ -0,0 +1,347 @@
+    It's ensured that the same character cannot appear two times in a row
+    chainLength: Expects an integer > 0 to determine the length of the string output
+    alphabet: A Python list containing all valid characters
+    """
+    output = ""
+    prevChar = ""
+    for _ in range(chainLength):
+        selectedChar = random.choice(alphabet)
+        while selectedChar==prevChar:
+            selectedChar = random.choice(alphabet)
+        output += selectedChar
+        prevChar = selectedChar
+    return output
+
+def generate_fibonacci(chainLength, alphabet):
+    match chainLength:
+        case 0:
+            return ""
+        case 1:
+            return random.choice(alphabet)
+
+    a = random.choice(alphabet)
+    b = random.choice(alphabet)
+    while a==b:
+        b = random.choice(alphabet)
+
+    while len(b)<chainLength:
+        a, b = b, a + b
+    return b[:chainLength]
+
+def randomRange(min, max):
+    if min >= max:
+        return min
+    else:
+        return random.choice(range(min,max))
+
+def generate_chains(nChains, alphabet, method, minLength, maxLength):
+    chains = []
+    
+    for i in range(nChains):
+        length = randomRange(minLength, maxLength)
+        chains.append(method(length, alphabet))
+    
+    return chains
+
+def border_array(x: str) -> list[int]:
+    """
+    Construct the border array for x.
+
+    >>> border_array("aaba")
+    [0, 1, 0, 1]
+    >>> border_array("ississippi")
+    [0, 0, 0, 1, 2, 3, 4, 0, 0, 1]
+    >>> border_array("")
+    []
+    >>> strict_border_array("abaabaa")
+    [0, 0, 1, 1, 2, 3, 4]
+    >>> border_array("abcabdabcabc")
+    [0, 0, 0, 1, 2, 0, 1, 2, 3, 4, 5, 3]
+    """
+    if x=="":
+        return []
+    border_list = [0 for _ in x]
+    for i, c in enumerate(x):
+        border_index = i
+        while True:
+            if border_index == 0:
+                # border_list[i] = 0 # The list was initialized with zeros
+                break
+            previousBorder = border_list[border_index-1]
+            if c == x[previousBorder]:
+                border_list[i] = previousBorder+1
+                break
+            # We go backwards throughout the border
+            border_index = previousBorder
+    return border_list
+
+# Returns the border length of the last border in the given string
+def lastBorder(x):
+    border = border_array(x)
+    index = len(border)-1
+    if index<0:
+        return 0
+    else:
+        return border[index]
+
+def findPattern(chain, pattern):
+    patternIndexes = []
+    start = 0
+    while True:
+        start = chain.find(pattern, start) + 1
+        if start > 0:
+            patternIndexes.append(start)
+        else:
+            return patternIndexes
+
+def embedString(base, insertion, index):
+    newChain = base[:index] + insertion + base[index+len(insertion):]
+    return newChain
+
+def __adapt_chain__(chain, pattern, min_matches, max_matches):
```

Comment:
Are you sure you want to make `__<dunder function>__` functions? The double underscore at the beginning and end of a method is usually reserved for Python. The double underscore at the beginning is a way to make a function hidden, but no one uses it. There isn't anything wrong with the name, but it is not what the usual Python programmer would expect.